lib/libzpool/kernel.c

```diff
@@ -117,7 +117,8 @@ zk_thread_create(void (*func)(void *), void *arg, size_t stksize, int state)
 	VERIFY0(pthread_attr_setstacksize(&attr, stksize));
 	VERIFY0(pthread_attr_setguardsize(&attr, PAGESIZE));
 
-	VERIFY0(pthread_create(&tid, &attr, (void *(*)(void *))func, arg));
+	VERIFY0(pthread_create(&tid, &attr,
+	    (void *(*)(void *))(uintptr_t)func, arg));
```

Comment:
Why cast to `(uintptr_t)`?  You're not doing bitwise operations on pointers.  Why not `(void *)`, if you need the cast at all?

What's the warning you're trying to silence?

Reply:
`func` is `void (*)(void *)`, `pthread_create()` takes `void *(*)(void *)`, this produces an invalid cast warning (and rightfully so):
```
In file included from kernel.c:27:
kernel.c: In function ‘zk_thread_create’:
kernel.c:121:6: warning: cast between incompatible function types from ‘void (*)(void *)’ to ‘void * (*)(void *)’ [-Wcast-function-type]
  121 |      (void *(*)(void *))func, arg));
      |      ^
../../lib/libspl/include/assert.h:104:37: note: in definition of macro ‘VERIFY0’
  104 |  const uint64_t __left = (uint64_t)(LEFT);   \
      |                                     ^~~~
```

What I did think about doing was allocating a (func, arg) bundle and pushing it through a helper, but this would screw with the (seemingly) precise stack size measurements, so.

I mean, casting it like this is obviously technically UB *anyway* because of the incompatible return types, but going through an `uintptr_t` (or, indeed, a pointer) is worse? But then save for a union or screwing with the stack size stuff I can't realistically come up with a good way of making the compiler go away here.

Reply:
Hmm.

- Converting between function pointers is permitted by the standard (you may not dereference (ie., call), but you can store).  See 6.3.2.3/8:

> A pointer to a function of one type may be converted to a pointer to a function of another type and
> back again; the result shall compare equal to the original pointer. If a converted pointer is used to
> call a function whose type is not compatible with the referenced type, the behavior is undefined.

- The call is being done in a different translation unit, and therefore, the compiler can't know if you will invoke UB or not, so it has to assume you're not invoking UB, and so it needs to follow the ABI.
- If the ABI defines both function pointers to be compatible (I assume they are, or you'd've noticed before), then, it's defined behavior ($DEITY bless translation unit boundaries to workaround UB).

So I'd say you're not invoking UB with the function pointer cast, but you need a cast.  I'd either use `-Wno-cast-function-type` (since the compiler is invalidly blaming you) or cast to `(void *)` before casting to the function pointer, if you need to keep that warning.  Choose your poison.

Reply:
Good find – I rolled back to the straight cast and added `-Wno-cast-function-type` to this file, pointing at this thread. It should™ be fine.

Reply:
BTW, there's another thing you could do to not disable the warning from the compiler: use `noreturn`.

As weird as it may seem, `noreturn int foo(void)` is valid C.

So you may use the prototype that `pthread_create(3)` expects, mark it as `noreturn`, and then the compiler should not warn about it.

Example:

```
alx@ady1:~/src/test$ cat noreturn.c
#include <pthread.h>
#include <stdnoreturn.h>

noreturn void *
foo(void *x)
{
	pthread_exit(x);
}

int
bar(pthread_t *restrict th, void *arg)
{
	return pthread_create(th, NULL, &foo, arg);
}
alx@ady1:~/src/test$ cc -Wall -Wextra -Werror -S noreturn.c 
alx@ady1:~/src/test$ echo $?
0
```

But you'd need to modify a few functions for that.